src/shaka_scheme/system/base/Bytevector.cpp

```diff
@@ -0,0 +1,74 @@
+#include "shaka_scheme/system/base/Bytevector.hpp"
+
+#include "shaka_scheme/system/exceptions/IndexOutOfBoundsException.hpp"
+
+#include <cctype>
+#include <cstring>
+#include <utility>
+
+namespace shaka {
+
+Bytevector::Bytevector(std::size_t k) :
+    arr(new unsigned char[k]),
+    size(k) {}
+
+
+Bytevector::Bytevector(std::size_t k, unsigned char byte) :
+    arr(new unsigned char[k]),
+    size(k) {
+  for (std::size_t i = 0; i < k; ++i) {
+    this->arr[i] = byte;
+  }
+}
+
+Bytevector::Bytevector(std::initializer_list<unsigned char> l) :
+    arr(new unsigned char[l.size()]),
+    size(l.size()) {
+  std::size_t i = 0;
+  for (const auto& it : l) {
+    this->arr[i] = it;
+    ++i;
+  }
+}
+
+Bytevector::Bytevector(const Bytevector& other) :
+    arr(new unsigned char[other.length()]),
+    size(other.length()) {
+  std::memcpy(this->arr, other.arr, sizeof(unsigned char)*other.length());
+}
+
+Bytevector::Bytevector(Bytevector&& other) :
+    arr(std::move(other.arr)),
+    size(std::move(other.size)) {
+  this->arr = nullptr;
+  this->size = 0;
+}
+
+Bytevector::~Bytevector() {
+  delete arr;
+}
+
+std::size_t Bytevector::length() const {
+  return this->size;
+}
+
+unsigned char& Bytevector::operator[](std::size_t k) {
+  if (k >= size) {

```

Comment:
Might want to check for negative indices as well

Reply:
Because it's `std::size_t`, the indices cannot ever be negative! I follow the convention set by [`libstdc++` with indices on `std::vector<T>`](https://gcc.gnu.org/onlinedocs/gcc-4.6.3/libstdc++/api/a01116_source.html) being of a `size_type` which I assume cannot be negative and is unsigned.